source/common/grpc/async_client_impl.cc

```diff
@@ -86,23 +86,21 @@ void AsyncStreamImpl::onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
       onTrailers(std::move(headers));
       return;
     }
-    streamError(Common::httpToGrpcStatus(http_response_status));
+    // Technically this should be
+    // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md
+    // as given by Common::httpToGrpcStatus(), but the Google gRPC client treats
+    // this as GrpcStatus::Canceled.
+    streamError(Status::GrpcStatus::Canceled);
     return;
   }
   if (end_stream) {
     onTrailers(std::move(headers));
     return;
   }
-  callbacks_.onReceiveInitialMetadata(std::move(headers));
 }
 
 void AsyncStreamImpl::onData(Buffer::Instance& data, bool end_stream) {
-  ASSERT(!remote_closed_);
-  if (end_stream) {
-    streamError(Status::GrpcStatus::Internal);
-    return;
-  }
-
+  trailers_only_ = false;
```

Comment:
This could be inaccurate. The gRPC response spec is `Response → (Response-Headers *Length-Prefixed-Message Trailers) / Trailers-Only` so it is possible that a non-trailers-only response with a headers followed by a trailers if an non-immediate error happens.

Reply:
Yeah, good point, I have some independent thread with the gRPC team on this specific scenario, I know how to fix this in `Grpc::AsyncClientImpl`, I don't know how to fix it in the Google gRPC integration though.